isis/src/control/objs/ControlNet/ControlNet.cpp

```diff
@@ -817,247 +742,67 @@ namespace Isis {
-      // trees for each node and check that they're disjoint, and thus able to
-      // be joined.
-      QString pointId = leastEdge->Parent()->GetId();
-      ControlVertex *pointVertex = forest[pointId];
-
-      QString serialNum = leastEdge->ControlSN()->getSerialNumber();
-      ControlVertex *nodeVertex = forest[serialNum];
-
-      // When the edge joins two different trees (i.e., they have different
-      // roots), then add the edge to the minimum spanning tree and join the
-      // trees into one.  Otherwise, we have formed a cycle and should thus
-      // discard the edge.
-      if (pointVertex->getRoot() != nodeVertex->getRoot()) {
-        ControlVertex::join(pointVertex, nodeVertex);
-        trees--;
-        minimumTree.insert(leastEdge);
-        uniquePoints[pointVertex->getPoint()]++;
-      }
-    }
-
-    // Prune edges that go from a graph node to a point, but not from that
-    // point to another graph node.  We care about image (graph node)
-    // connectivity, not point connectivity.  A complete edge consists of two
-    // measures and a point, so remove any incomplete edges.
-    QList< ControlMeasure * > unprunedEdges = minimumTree.values();
-    for (int i = 0; i < unprunedEdges.size(); i++) {
-      if (uniquePoints[unprunedEdges[i]->Parent()] < 2)
-        // The point this edge is connected to does not go on to another node,
-        // so prune it
-        minimumTree.remove(unprunedEdges[i]);
-    }
-
-    // Clean up our vertices.  This will not delete any of the points, measures,
-    // or graph nodes.  All of that is owned by the network.
-    QList< ControlVertex * > vertexList = forest.values();
-    for (int i = 0; i < vertexList.size(); i++) delete vertexList[i];
-
-    // Sanity check: an island with n > 1 nodes must, by definition, have an MST
-    // of e edges such that n <= e <= 2n
-    int n = island.size();
-    int e = minimumTree.size();
-    if (n > 1) {
-      if (e < n || e > 2 * n) {
-        IString msg = "An island of n = [" + IString(n) +
-          "] > 1 nodes must have a minimum spanning tree of e edges such that "
-          " n <= e <= 2n, but e = [" + IString(e) + "]";
-        throw IException(IException::Programmer, msg, _FILEINFO_);
+    QList< QString> serials = m_vertexMap.keys(); 
+    for (int i = 0; i < serials.size(); i++) {
+      boost::put(indexMapAdaptor, m_vertexMap[serials[i]], i);
+    }
+
+    VertexIndexMap componentMap;
+    VertexIndexMapAdaptor componentAdaptor(componentMap);
+    size_t numComponents = boost::connected_components(m_controlGraph, componentAdaptor,
+                                                      boost::vertex_index_map(indexMapAdaptor));
+
+//    std::cout << "Number of components:" << std::endl;
+//    std::cout << numComponents << std::endl; 
+//    std::cout << "Size of return: " << componentMap.size() << std::endl; 
+
+    // This is not efficient, but works
+    QHash<int, QList<QString> > islandsByGroup; 
```

Comment:
You know how many islands you have at this point, so you can create the output list right here.

Reply:
This way you don't need a second loop

Reply:
:+1: Thank you. I even knew this was the wrong way to do it, I was just blanking on what the right way was.